File: 2448-minimum-cost-to-make-array-equal/2448-minimum-cost-to-make-array-equal.java
class Solution {
    class Pair implements Comparable<Pair>{
        int num;
        int cost;
        
        public int compareTo(Pair o){
            return this.num-o.num;
        }
    }
    public long minCost(int[] nums, int[] cost) {
        Pair[] a=new Pair[nums.length];
        int n =nums.length;
        for(int i=0;i<nums.length;i++){
            Pair p=new Pair();
            p.num=nums[i];
            p.cost=cost[i];
            a[i]=p;
        }
        Arrays.sort(a);
        
        
        long[] costL=new long[n];
        costL[0]=a[0].cost;
        for(int i=1;i<n;i++){
            costL[i]=a[i].cost + costL[i-1];
        }
        
        long total = 0;
        for(int i = 1; i < n; i++){
            total += 1l*  a[i].cost * (a[i].num - a[0].num);
        }
        
        long ans = total ; 
        for(int i=1;i<n;i++){
            int gap = a[i].num - a[i-1].num;
            total += 1l * costL[i - 1] * gap;
            total -= 1l * (costL[n-1] - costL[i-1]) * gap;
            
            ans = Math.min(total, ans);
        }
        return ans;
    }
}

File: 2448-minimum-cost-to-make-array-equal/2448-minimum-cost-to-make-array-equal.cpp
class Solution {
public:
    long long minCost(vector<int>& n, vector<int>& cost) {
        vector<long long> id(n.size()), cost_l(n.size());
        iota(begin(id), end(id), 0);
        sort(begin(id), end(id), [&](int i, int j){
            return n[i] < n[j];
        });
        for (long long i = 0, psum = 0; i < n.size() - 1; ++i) {
            psum += cost[id[i]];
            cost_l[i + 1] = cost_l[i] + psum * (n[id[i + 1]] - n[id[i]]);
        }
        long long res = cost_l.back(), cost_r = 0;
        for (long long j = n.size() - 1, psum = 0; j > 0; --j) {
            psum += cost[id[j]];
            cost_r += psum * (n[id[j]] - n[id[j - 1]]);
            res = min(res, cost_l[j - 1] + cost_r);
        }
        return res;
    }
};